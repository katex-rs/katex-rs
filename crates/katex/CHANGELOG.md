# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0](https://github.com/katex-rs/katex-rs/compare/katex-rs-v0.1.1...katex-rs-v0.2.0) - 2025-09-30

### Other

- Refactor data extraction scripts to Rust
- Update GitHub CI workflow for the latest features and configure release-plz
- Split wasm binding to a seperated crate
- Better token processing
- Use Arc in function and env map
- Reduce clone in build_html
- Simplify CssStyle for most cases
- Avoid vec flatten with push_combine_chars and optimize the parser by removing unnecessary clones
- Use ClassList to avoid clones in class list manipulations
- Make various optimizations
- Add flamegraph xtask tool
- Fix svg_geometry
- Refactor the project to use cargo  workspace
