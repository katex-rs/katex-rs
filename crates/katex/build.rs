extern crate alloc;

use alloc::collections::{BTreeMap, BTreeSet};
use core::{
    error::Error as CoreError,
    fmt::{self, Write as _},
    iter::Peekable,
};
use std::{
    env, fs,
    fs::File,
    io::{BufWriter, Write as _},
    path::PathBuf,
};

type BuildResult<T> = Result<T, Box<dyn CoreError>>;

#[derive(Debug)]
struct BuildScriptError(String);

impl fmt::Display for BuildScriptError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

impl CoreError for BuildScriptError {}

#[path = "src/unicode/unicode_accents.rs"]
mod unicode_accents;
use unicode_accents::UNICODE_ACCENTS;

#[derive(serde::Deserialize)]
struct Symbol {
    mode: String,
    font: String,
    group: String,
    replace: Option<String>,
    name: String,
    #[serde(rename = "acceptUnicodeChar")]
    accept_unicode_char: bool,
}

fn main() -> BuildResult<()> {
    println!("cargo:rerun-if-changed=data/font_metrics_data.json");
    println!("cargo:rerun-if-changed=data/symbols.json");
    println!("cargo:rerun-if-changed=data/sigmas_and_xis.json");

    let out_dir = PathBuf::from(env::var("OUT_DIR")?);

    let sigmas = generate_sigmas_and_xis()?;
    write_file(out_dir.join("sigmas_and_xis_generated.rs"), &sigmas)?;

    let font_metrics = generate_font_metrics()?;
    write_file(out_dir.join("font_metrics_data_phf.rs"), &font_metrics)?;

    let unicode_symbols = generate_unicode_symbols()?;
    write_file(out_dir.join("unicode_symbols_phf.rs"), &unicode_symbols)?;

    let symbols = generate_symbols()?;
    write_file(out_dir.join("generated_symbols_data.rs"), &symbols)?;

    Ok(())
}

// Function to convert camelCase to snake_case
fn to_snake_case(s: &str) -> String {
    let mut result = String::new();
    for (i, ch) in s.chars().enumerate() {
        if ch.is_uppercase() && i > 0 {
            result.push('_');
        }
        if let Some(lower) = ch.to_lowercase().next() {
            result.push(lower);
        }
    }
    result
}

fn generate_sigmas_and_xis() -> BuildResult<String> {
    let data = read_json("data/sigmas_and_xis.json")?;
    let sigmas_and_xis = data
        .get("sigmasAndXis")
        .and_then(|value| value.as_object())
        .ok_or_else(|| BuildScriptError("sigmasAndXis object".into()))?;
    let field_docs = data
        .get("fieldDocs")
        .and_then(|value| value.as_object())
        .ok_or_else(|| BuildScriptError("fieldDocs object".into()))?;

    let mut fields = String::new();
    for (key, _) in sigmas_and_xis {
        let snake_key = to_snake_case(key);
        if let Some(doc) = field_docs.get(key).and_then(|value| value.as_str()) {
            let _ = writeln!(&mut fields, "    /// {doc}");
        }
        let _ = writeln!(&mut fields, "    pub {snake_key}: f64,");
    }
    let _ = writeln!(
        &mut fields,
        "    /// CSS em per mu\n    pub css_em_per_mu: f64,"
    );

    let mut consts = String::new();
    let quad = sigmas_and_xis
        .get("quad")
        .and_then(|value| value.as_array())
        .ok_or_else(|| BuildScriptError("quad should be an array".into()))?;

    for index in 0..3 {
        let _ = writeln!(&mut consts, "    FontMetrics {{");
        for (key, value) in sigmas_and_xis {
            let Some(values) = value.as_array() else {
                continue;
            };
            let Some(entry) = values.get(index) else {
                continue;
            };
            let formatted = if entry.is_i64() {
                entry
                    .as_f64()
                    .map_or_else(|| entry.to_string(), |value| format!("{value:.1}"))
            } else {
                entry.to_string()
            };
            let snake = to_snake_case(key);
            let _ = writeln!(&mut consts, "        {snake}: {formatted},");
        }

        let css_em_per_mu = quad
            .get(index)
            .and_then(serde_json::Value::as_f64)
            .ok_or_else(|| BuildScriptError("quad entry should be a float".into()))?
            / 18.0;
        let _ = writeln!(&mut consts, "        css_em_per_mu: {css_em_per_mu},");
        let _ = writeln!(&mut consts, "    }},");
    }

    Ok(format!(
        "// Auto-generated from KaTeX/src/fontMetrics.js
// Do not edit manually

#[derive(Debug, Clone)]
/// Font metrics for a specific style size (text, script, scriptscript)
pub struct FontMetrics {{
{fields}}}

/// Constant font metrics for textstyle, scriptstyle, and scriptscriptstyle
pub const FONT_METRICS: [FontMetrics; 3] = [
{consts}];
"
    ))
}

fn generate_font_metrics() -> BuildResult<String> {
    use phf_codegen::Map as PhfMap;

    let json_data = fs::read_to_string("data/font_metrics_data.json")?;
    let font_metrics: BTreeMap<String, BTreeMap<String, Vec<f64>>> =
        serde_json::from_str(&json_data)?;

    let mut output = String::new();
    let mut font_index = PhfMap::new();

    for (font_family, metrics) in font_metrics {
        let font_name = font_family.replace(['-', '.'], "_");
        font_index.entry(
            font_family.clone(),
            format!("&{}_METRICS", font_name.to_uppercase()),
        );

        let mut map = PhfMap::new();
        for (char_code_str, metrics_array) in metrics {
            let char_code: u32 = char_code_str.parse()?;
            let values = metrics_array
                .iter()
                .map(|v| {
                    if v.fract() == 0.0 {
                        format!("{v:.1}")
                    } else {
                        v.to_string()
                    }
                })
                .collect::<Vec<_>>()
                .join(", ");
            map.entry(char_code, format!("CharacterMetrics::new({values})"));
        }

        let _ = writeln!(
            &mut output,
            "/// Font metrics for the {font_family} font family"
        );
        let _ = writeln!(
            &mut output,
            "#[allow(clippy::expect_used)]\n#[allow(clippy::approx_constant)]\npub const {}_METRICS: phf::Map<u32, CharacterMetrics> = {};\n",
            font_name.to_uppercase(),
            map.build()
        );
    }

    let _ = writeln!(
        &mut output,
        "/// Mapping of font family names to their corresponding metrics maps"
    );
    let _ = writeln!(
        &mut output,
        "#[allow(clippy::expect_used)]\n#[allow(clippy::non_ascii_literal)]\npub const FONT_METRICS_INDEX: phf::Map<&'static str, &'static phf::Map<u32, CharacterMetrics>> = \n{};\n",
        font_index.build()
    );

    Ok(output)
}

#[allow(clippy::unnecessary_wraps)]
fn generate_unicode_symbols() -> BuildResult<String> {
    use phf_codegen::Map as PhfMap;

    let accents: Vec<char> = UNICODE_ACCENTS.keys().copied().collect();
    let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ\u{3b1}\u{3b2}\u{3b3}\u{3b4}\u{3b5}\u{3f5}\u{3b6}\u{3b7}\u{3b8}\u{3d1}\u{3b9}\u{3ba}\u{3bb}\u{3bc}\u{3bd}\u{3be}\u{3bf}\u{3c0}\u{3d6}\u{3c1}\u{3f1}\u{3c2}\u{3c3}\u{3c4}\u{3c5}\u{3c6}\u{3d5}\u{3c7}\u{3c8}\u{3c9}\u{393}\u{394}\u{398}\u{39b}\u{39e}\u{3a0}\u{3a3}\u{3a5}\u{3a6}\u{3a8}\u{3a9}";

    let mut seen = BTreeSet::new();
    let mut map = PhfMap::<char>::new();

    for letter in letters.chars() {
        for accent in &accents {
            let combined = format!("{letter}{accent}");
            if let Some(normalized) = normalize_to_single_char(&combined)
                && seen.insert(normalized)
            {
                map.entry(normalized, format!("\"{}\"", escape_as_unicode(&combined)));
            }

            for accent2 in &accents {
                if accent == accent2 {
                    continue;
                }

                let combined2 = format!("{letter}{accent2}{accent}");
                if let Some(normalized) = normalize_to_single_char(&combined2)
                    && seen.insert(normalized)
                {
                    map.entry(normalized, format!("\"{}\"", escape_as_unicode(&combined2)));
                }
            }
        }
    }

    Ok(format!(
        "/// Mapping of normalized Unicode symbols to their component parts
/// Unicode symbols map for Modifier tone letters
#[allow(clippy::expect_used)]
#[allow(clippy::non_ascii_literal)]
pub const UNICODE_SYMBOLS: phf::Map<char, &str> = \n{};\n",
        map.build()
    ))
}

fn generate_symbols() -> BuildResult<String> {
    let json_data = fs::read_to_string("data/symbols.json")?;
    let symbol_data: Vec<Symbol> = serde_json::from_str(&json_data)?;

    let (math_symbols, text_symbols): (Vec<Symbol>, Vec<Symbol>) =
        symbol_data.into_iter().partition(|s| s.mode == "math");

    let mut output = String::new();
    let _ = writeln!(&mut output, "// Auto-generated file - do not edit manually");
    let _ = writeln!(&mut output, "// Generated from data/symbols.json\n");
    let _ = writeln!(&mut output, "/// Populate math symbols from JSON data");
    write_symbols(&mut output, &math_symbols, "POPULATE_MATH_SYMBOLS")?;
    let _ = writeln!(&mut output, "/// Populate text symbols from JSON data");
    write_symbols(&mut output, &text_symbols, "POPULATE_TEXT_SYMBOLS")?;

    Ok(output)
}

fn write_file(path: PathBuf, contents: &str) -> BuildResult<()> {
    let mut writer = BufWriter::new(File::create(path)?);
    writer.write_all(contents.as_bytes())?;
    Ok(())
}

fn read_json(path: &str) -> BuildResult<serde_json::Value> {
    let contents = fs::read_to_string(path)?;
    Ok(serde_json::from_str(&contents)?)
}

fn escape_as_unicode(s: &str) -> String {
    s.chars()
        .map(|c| {
            if c.is_ascii() && !c.is_control() {
                c.to_string()
            } else {
                format!("\\u{{{:x}}}", c as u32)
            }
        })
        .collect::<String>()
}

fn parse_unicode_escape<I>(iter: &mut Peekable<I>) -> Option<String>
where
    I: Iterator<Item = char> + Clone,
{
    let mut lookahead = iter.clone();
    if lookahead.next()? != '\\' {
        return None;
    }
    if lookahead.next()? != 'u' {
        return None;
    }

    if matches!(lookahead.peek(), Some('{')) {
        lookahead.next();
        let mut hex = String::new();
        while let Some(&c) = lookahead.peek() {
            if c == '}' {
                break;
            }
            if c.is_ascii_hexdigit() {
                hex.push(c);
                lookahead.next();
            } else {
                return None;
            }
        }
        if hex.is_empty() || lookahead.next()? != '}' {
            return None;
        }
        *iter = lookahead;
        return Some(format!("\\u{{{hex}}}"));
    }

    let mut hex = String::new();
    for _ in 0..4 {
        let c = lookahead.next()?;
        if c.is_ascii_hexdigit() {
            hex.push(c);
        } else {
            return None;
        }
    }
    *iter = lookahead;
    Some(format!("\\u{{{hex}}}"))
}

fn convert_unicode_escapes(s: &str) -> String {
    let mut result = String::new();
    let mut chars = s.chars().peekable();
    while chars.peek().is_some() {
        if let Some(escaped) = parse_unicode_escape(&mut chars) {
            result.push_str(&escaped);
        } else if let Some(next) = chars.next() {
            result.push(next);
        } else {
            break;
        }
    }
    result
}

fn normalize_to_single_char(s: &str) -> Option<char> {
    use unicode_normalization::UnicodeNormalization as _;

    let normalized: String = s.chars().nfc().collect();
    let mut chars = normalized.chars();
    match (chars.next(), chars.next()) {
        (Some(c), None) => Some(c),
        _ => None,
    }
}

fn write_symbols(buffer: &mut String, symbols: &[Symbol], variant: &str) -> BuildResult<()> {
    let mut grouped: BTreeMap<(String, String, Option<String>), BTreeSet<String>> = BTreeMap::new();

    for symbol in symbols {
        let key = (
            symbol.font.clone(),
            symbol.group.clone(),
            symbol.replace.clone(),
        );
        let entry = grouped.entry(key).or_default();
        entry.insert(symbol.name.clone());

        if let (true, Some(replace)) = (symbol.accept_unicode_char, &symbol.replace) {
            entry.insert(replace.clone());
        }
    }

    let _ = writeln!(
        buffer,
        "const {variant}_MAP: phf::Map<&str, CharInfo> = phf_map!("
    );

    for ((font, group, replace), names) in grouped {
        let arm = names
            .into_iter()
            .map(|s| convert_unicode_escapes(&s))
            .collect::<Vec<_>>()
            .join("\" | \"");

        let font_str = match font.as_str() {
            "main" => "Font::Main",
            "ams" => "Font::Ams",
            _ => return Err(BuildScriptError(format!("Invalid font: {font}")).into()),
        };

        let group_str = match group.as_str() {
            "bin" => "Group::Atom(Atom::Bin)",
            "close" => "Group::Atom(Atom::Close)",
            "inner" => "Group::Atom(Atom::Inner)",
            "open" => "Group::Atom(Atom::Open)",
            "punct" => "Group::Atom(Atom::Punct)",
            "rel" => "Group::Atom(Atom::Rel)",
            "accent" | "accent-token" => "Group::NonAtom(NonAtom::AccentToken)",
            "mathord" => "Group::NonAtom(NonAtom::MathOrd)",
            "op" | "op-token" => "Group::NonAtom(NonAtom::OpToken)",
            "spacing" => "Group::NonAtom(NonAtom::Spacing)",
            "textord" => "Group::NonAtom(NonAtom::TextOrd)",
            _ => return Err(BuildScriptError(format!("Invalid group: {group}")).into()),
        };

        let replace_value = replace.as_ref().map_or_else(
            || "None".to_owned(),
            |s| format!("Some(\'{}\')", convert_unicode_escapes(s)),
        );

        let _ = writeln!(buffer, "    \"{arm}\" => CharInfo {{");
        let _ = writeln!(buffer, "        font: {font_str},");
        let _ = writeln!(buffer, "        group: {group_str},");
        let _ = writeln!(buffer, "        replace: {replace_value},");
        let _ = writeln!(buffer, "    }},");
    }

    let _ = writeln!(buffer, ");\n");

    Ok(())
}
