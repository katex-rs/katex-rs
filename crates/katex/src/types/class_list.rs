//! Abstraction for managing CSS class lists across DOM nodes.

use alloc::borrow::Cow;
use core::slice;

/// Iterator over class names stored in a [`ClassList`].
pub enum ClassListIter<'a> {
    /// Iterator over owned class storage.
    Owned(slice::Iter<'a, Cow<'static, str>>),
    /// Iterator yielding a single static class.
    Static(Option<&'a str>),
    /// Iterator over a compile-time set of static classes.
    Const(slice::Iter<'a, &'static str>),
    /// Empty iterator.
    Empty,
}

/// Represents a collection of CSS classes applied to a DOM node.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub enum ClassList {
    /// Owned storage for the list of classes.
    Owned(Vec<Cow<'static, str>>),
    /// Single static class stored without allocation.
    Static(&'static str),
    /// Multiple static classes stored without allocation.
    Const(&'static [&'static str]),
    /// Explicit empty list of classes.
    #[default]
    Empty,
}

const CLASS_PREALLOCATE: usize = 4;

impl ClassList {
    /// Returns `true` if the list contains no classes.
    #[must_use]
    pub const fn is_empty(&self) -> bool {
        match self {
            Self::Owned(vec) => vec.is_empty(),
            Self::Static(_) => false,
            Self::Const(slice) => slice.is_empty(),
            Self::Empty => true,
        }
    }

    /// Returns the number of classes in the list.
    #[must_use]
    pub const fn len(&self) -> usize {
        match self {
            Self::Owned(vec) => vec.len(),
            Self::Static(_) => 1,
            Self::Const(slice) => slice.len(),
            Self::Empty => 0,
        }
    }

    /// Returns the first class if present.
    #[inline]
    #[must_use]
    pub fn first(&self) -> Option<&str> {
        match self {
            Self::Owned(vec) => vec.first().map(AsRef::as_ref),
            Self::Static(class) => Some(*class),
            Self::Const(slice) => Some(*slice.first()?),
            Self::Empty => None,
        }
    }

    /// Gets a class at a specific index if it exists.
    #[must_use]
    pub fn get(&self, index: usize) -> Option<&str> {
        match self {
            Self::Owned(vec) => vec.get(index).map(AsRef::as_ref),
            Self::Static(class) => Some(*class),
            Self::Const(slice) => slice.get(index).map(AsRef::as_ref),
            Self::Empty => None,
        }
    }

    /// Gets a mutable reference to a class at a specific index.
    pub fn get_mut(&mut self, index: usize) -> Option<&mut Cow<'static, str>> {
        let vec = self.ensure_owned();
        vec.get_mut(index)
    }

    /// Reserves capacity for additional classes when owned.
    pub fn reserve(&mut self, additional: usize) {
        if additional == 0 {
            return;
        }
        let vec = self.ensure_owned();
        vec.reserve(additional);
    }

    /// Returns `true` if the class list contains the specified class name.
    #[must_use]
    pub fn contains(&self, class: &str) -> bool {
        match self {
            Self::Owned(vec) => vec.iter().any(|entry| entry == class),
            Self::Static(entry) => *entry == class,
            Self::Const(entries) => entries.contains(&class),
            Self::Empty => false,
        }
    }

    #[inline]
    fn ensure_owned(&mut self) -> &mut Vec<Cow<'static, str>> {
        match self {
            Self::Owned(v) => v,
            _ => self.inner_owned(),
        }
    }

    #[cold]
    fn inner_owned(&mut self) -> &mut Vec<Cow<'static, str>> {
        let v = match self {
            Self::Empty => Vec::with_capacity(CLASS_PREALLOCATE),
            Self::Static(class) => {
                let mut v = Vec::with_capacity(CLASS_PREALLOCATE);
                if !class.is_empty() {
                    v.push(Cow::Borrowed(*class));
                }
                v
            }
            Self::Const(slice) => {
                let mut v = Vec::with_capacity(slice.len().saturating_add(CLASS_PREALLOCATE));
                v.extend(slice.iter().map(|&s| Cow::Borrowed(s)));
                v
            }
            Self::Owned(_) => unreachable!(),
        };

        *self = Self::Owned(v);
        if let Self::Owned(v) = self {
            v
        } else {
            unreachable!()
        }
    }

    /// Pushes a class onto the list, ignoring empty entries.
    pub fn push<S>(&mut self, class: S)
    where
        S: Into<Cow<'static, str>>,
    {
        let class = class.into();
        if class.is_empty() {
            return;
        }

        let vec = self.ensure_owned();
        vec.push(class);
    }

    /// Inserts a class at the specified position, ignoring empty entries.
    pub fn insert<S>(&mut self, index: usize, class: S)
    where
        S: Into<Cow<'static, str>>,
    {
        let class = class.into();
        if class.is_empty() {
            return;
        }

        let vec = self.ensure_owned();
        vec.insert(index, class);
    }

    /// Iterate over all classes in the list.
    #[must_use]
    pub fn iter(&self) -> ClassListIter<'_> {
        match self {
            Self::Owned(vec) => ClassListIter::Owned(vec.iter()),
            Self::Static(class) => {
                if class.is_empty() {
                    ClassListIter::Empty
                } else {
                    ClassListIter::Static(Some(*class))
                }
            }
            Self::Const(slice) => ClassListIter::Const(slice.iter()),
            Self::Empty => ClassListIter::Empty,
        }
    }
}

impl From<Vec<Cow<'static, str>>> for ClassList {
    fn from(value: Vec<Cow<'static, str>>) -> Self {
        Self::Owned(value)
    }
}

impl<'a> IntoIterator for &'a ClassList {
    type Item = &'a str;
    type IntoIter = ClassListIter<'a>;

    fn into_iter(self) -> Self::IntoIter {
        self.iter()
    }
}

impl Extend<Cow<'static, str>> for ClassList {
    fn extend<T: IntoIterator<Item = Cow<'static, str>>>(&mut self, iter: T) {
        let vec = self.ensure_owned();
        vec.extend(iter.into_iter().filter(|s| !s.is_empty()));
    }
}

impl From<&'static str> for ClassList {
    fn from(value: &'static str) -> Self {
        Self::Static(value)
    }
}

impl From<&'static [&'static str]> for ClassList {
    fn from(value: &'static [&'static str]) -> Self {
        Self::Const(value)
    }
}

impl<'a> Iterator for ClassListIter<'a> {
    type Item = &'a str;

    fn next(&mut self) -> Option<Self::Item> {
        match self {
            ClassListIter::Owned(iter) => iter.next().map(AsRef::as_ref),
            ClassListIter::Static(class) => class.take(),
            ClassListIter::Const(iter) => {
                for class in iter {
                    if !class.is_empty() {
                        return Some(*class);
                    }
                }
                None
            }
            ClassListIter::Empty => None,
        }
    }
}
