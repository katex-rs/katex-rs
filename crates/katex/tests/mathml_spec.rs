#![allow(clippy::non_ascii_literal)]

mod setup;

use katex::types::{Settings, StrictSetting, TrustSetting};
use setup::*;

#[test]
fn a_mathml_builder() {
    it("should generate the right types of nodes", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\sin{x}+1\;\text{a}", &settings)?;
        insta::assert_snapshot!("mathml_spec__right_types_of_nodes", markup);
        Ok(())
    });

    it("should concatenate digits into single <mn>", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\sin{\alpha}=0.34=.34^1", &settings)?;
        insta::assert_snapshot!("mathml_spec__concatenate_digits", markup);
        let markup = mathml_markup(r"1{,}000{,}000", &settings)?;
        insta::assert_snapshot!("mathml_spec__concatenate_digits_commas", markup);
        Ok(())
    });

    it("should make prime operators into <mo> nodes", || {
        let settings = Settings::default();
        let markup = mathml_markup("f'", &settings)?;
        insta::assert_snapshot!("mathml_spec__prime_operator", markup);
        Ok(())
    });

    it("should generate <mphantom> nodes for \\phantom", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\phantom{x}", &settings)?;
        insta::assert_snapshot!("mathml_spec__phantom", markup);
        Ok(())
    });

    it("should use <munderover> for large operators", || {
        let mut settings = Settings::default();
        settings.display_mode = true;
        let markup = mathml_markup(r"\displaystyle\sum_a^b", &settings)?;
        insta::assert_snapshot!("mathml_spec__munderover_large_operators", markup);
        Ok(())
    });

    it("should use <msupsub> for integrals", || {
        let mut settings = Settings::default();
        settings.display_mode = true;
        let markup = mathml_markup(
            r"\displaystyle\int_a^b + \oiint_a^b + \oiiint_a^b",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__msupsub_integrals", markup);
        Ok(())
    });

    it("should use <msupsub> for regular operators", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\textstyle\sum_a^b", &settings)?;
        insta::assert_snapshot!("mathml_spec__msupsub_regular_operators", markup);
        Ok(())
    });

    it(
        "should output \\limsup_{x \\rightarrow \\infty} correctly in \\textstyle",
        || {
            let settings = Settings::default();
            let markup = mathml_markup(r"\limsup_{x \\rightarrow \\infty}", &settings)?;
            insta::assert_snapshot!("mathml_spec__limsup_textstyle", markup);
            Ok(())
        },
    );

    it(
        "should output \\limsup_{x \\rightarrow \\infty} in displaymode correctly",
        || {
            let settings = Settings::builder().display_mode(true).build();
            let markup = mathml_markup(r"\limsup_{x \\rightarrow \\infty}", &settings)?;
            insta::assert_snapshot!("mathml_spec__limsup_displaymode", markup);
            Ok(())
        },
    );

    it("should use <mpadded> for raisebox", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\raisebox{0.25em}{b}", &settings)?;
        insta::assert_snapshot!("mathml_spec__raisebox", markup);
        Ok(())
    });

    it("should size delimiters correctly", || {
        let settings = Settings::default();
        let markup = mathml_markup(
            r"(M) \\big(M\\big) \\Big(M\\Big) \\bigg(M\\bigg) \\Bigg(M\\Bigg)",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__sized_delimiters", markup);
        Ok(())
    });

    it("should use <menclose> for colorbox", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\colorbox{red}{b}", &settings)?;
        insta::assert_snapshot!("mathml_spec__menclose_colorbox", markup);
        Ok(())
    });

    it("should build the CD environment properly", || {
        let settings = Settings::builder()
            .display_mode(true)
            .strict(StrictSetting::Bool(false))
            .build();
        let markup = mathml_markup(
            r"\begin{CD} A @>a>> B\\\\ @VVbV @VVcV\\\\ C @>d>> D \end{CD}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__cd_environment", markup);
        Ok(())
    });

    it("should set href attribute for href appropriately", || {
        let settings = Settings::builder().trust(TrustSetting::Bool(true)).build();
        let markup = mathml_markup(r"\href{http://example.org}{\alpha}", &settings)?;
        insta::assert_snapshot!("mathml_spec__href_trusted", markup);

        let default_settings = Settings::default();
        let _ = mathml_markup(
            r"p \Vdash \beta \href{http://example.org}{+ \alpha} \times \gamma",
            &default_settings,
        )?;
        Ok(())
    });

    it("should render mathchoice as if there was nothing", || {
        let settings = Settings::default();
        let markup = mathml_markup(
            r"\displaystyle\mathchoice{\sum_{k = 0}^{\infty} x^k}{T}{S}{SS}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__mathchoice_display", markup);
        let markup = mathml_markup(
            r"\mathchoice{D}{\sum_{k = 0}^{\infty} x^k}{S}{SS}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__mathchoice_inline", markup);
        let markup = mathml_markup(
            r"x_{\mathchoice{D}{T}{\sum_{k = 0}^{\infty} x^k}{SS}}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__mathchoice_subscript", markup);
        let markup = mathml_markup(
            r"x_{y_{\mathchoice{D}{T}{S}{\sum_{k = 0}^{\infty} x^k}}}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__mathchoice_nested", markup);
        Ok(())
    });

    it(
        "should render boldsymbol with the correct mathvariants",
        || {
            let settings = Settings::default();
            let markup = mathml_markup(r"\boldsymbol{Ax2k\omega\Omega\imath+}", &settings)?;
            insta::assert_snapshot!("mathml_spec__boldsymbol_mathvariants", markup);
            Ok(())
        },
    );

    it(
        "accents turn into <mover accent=\"true\"> in MathML",
        || {
            let settings = Settings::builder()
                .strict(StrictSetting::Bool(false))
                .build();
            let markup = mathml_markup("über fiancée", &settings)?;
            insta::assert_snapshot!("mathml_spec__accent_mover", markup);
            Ok(())
        },
    );

    it("tags use <mlabeledtr>", || {
        let settings = Settings::builder().display_mode(true).build();
        let markup = mathml_markup(r"\tag{hi} x+y^2", &settings)?;
        insta::assert_snapshot!("mathml_spec__tag_mlabeledtr", markup);
        Ok(())
    });

    it("normal spaces render normally", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\kern1em\kern1ex", &settings)?;
        insta::assert_snapshot!("mathml_spec__normal_spaces", markup);
        Ok(())
    });

    it("special spaces render specially", || {
        let settings = Settings::default();
        let markup = mathml_markup(
            r"\,\thinspace\:\>\medspace\;\thickspace\!\negthinspace\negmedspace\negthickspace\mkern1mu\mkern3mu\mkern4mu\mkern5mu\mkern-1mu\mkern-3mu\mkern-4mu\mkern-5mu",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__special_spaces", markup);
        Ok(())
    });

    it("ligatures render properly", || {
        let settings = Settings::default();
        let markup = mathml_markup(
            r"\text{```Hi----'''}`--\texttt{```Hi----'''}`\text{\tt ```Hi----'''}`",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__ligatures", markup);
        Ok(())
    });

    it("\\text fonts become mathvariant", || {
        let settings = Settings::default();
        let markup = mathml_markup(
            r"\text{roman\textit{italic\textbf{bold italic}}\textbf{bold}\textsf{ss\textit{italic\textbf{bold italic}}\textbf{bold}}\texttt{tt\textit{italic\textbf{bold italic}}\textbf{bold}}}",
            &settings,
        )?;
        insta::assert_snapshot!("mathml_spec__text_fonts_mathvariant", markup);
        Ok(())
    });

    it("\\html@mathml makes clean symbols", || {
        let settings = Settings::default();
        let markup = mathml_markup(r"\copyright\neq\notin\u2258\KaTeX", &settings)?;
        insta::assert_snapshot!("mathml_spec__htmlmathml_clean_symbols", markup);
        Ok(())
    });
}
