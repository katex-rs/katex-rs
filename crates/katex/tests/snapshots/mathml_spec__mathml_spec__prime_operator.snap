---
source: crates/katex/tests/mathml_spec.rs
assertion_line: 29
expression: markup
---
<math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>f</mi><mo lspace="0em" mathvariant="normal" rspace="0em">â€²</mo></msup></mrow><annotation encoding="application/x-tex">f&#x27;</annotation></semantics></math>
